File: exs/c/gethttp.c
//
// gethttp.c
//
// Retrieves a file using the Hyper Text Transfer Protocol
// and prints its contents to stdout.
//

//
// Pass the server name and full path of the file on the
// command line and redirect the output to a file. The program
// prints messages to stderr as it progresses.
//
// Example:
//		GetHTTP www.idgbooks.com /index.html > index.html
//

#include <stdio.h>
#include <string.h>
#include <io.h>
#include <fcntl.h>
#include "..\..\inc\dsock.h"

void GetHTTP( char far * lpServerName, char far * lpFileName );

// Helper macro for displaying errors
#define PRINTERROR( s )	\
		fprintf( stderr,"\n%s: %d\n", s, WSAGetLastError( ) )


///
static long near clock ( void )
{
    asm     push    es
    asm     xor     ax, ax
    asm     mov     es, ax
    asm     mov     ax, es:[46Ch+0]
    asm     mov     dx, es:[46Ch+2]
    asm     pop     es
}

///////////////////////////////////////////////////////////////////////////
void main( int argc, char *argv[] )
{
	WORD wVersionRequested = MAKEWORD( 1,1 );
	WSADATA wsaData;
	int nRet;

    //
	// Check arguments
	//
	if ( argc != 3 )
	{
		fprintf( stderr,
				 "\nSyntax: GetHTTP ServerName FullPathName\n" );
		return;
	}

	//
	// Initialize WinSock.dll
	//
	nRet = WSAStartup( wVersionRequested, &wsaData );
	if ( nRet )
	{
		fprintf( stderr,"\nWSAStartup( ): %d\n", nRet );
		WSACleanup( );
		return;
	}

	//
	// Check WinSock version
	//
	if ( wsaData.wVersion != wVersionRequested )
	{
		fprintf( stderr,"\nWinSock version not supported\n" );
		WSACleanup( );
		return;
	}

	fprintf( stderr, "\nWinSock ver:%d.%d desc:%s\n",
			 wsaData.wVersion >> 8, wsaData.wVersion & 0xFF,
			 wsaData.szDescription );

	//
	// Set "stdout" to binary mode
	//
	setmode( fileno( stdout), O_BINARY );

	//
	// Call GetHTTP( ) to do all the work
	//
	GetHTTP( argv[1], argv[2] );

	//
	// Release WinSock
	//
	fprintf( stderr, "\nFinishing..." );
	WSACleanup( );
	fprintf( stderr, " done.\n" );
}

///////////////////////////////////////////////////////////////////////////
void GetHTTP( char far *lpServerName, char far *lpFileName )
{
	IN_ADDR     iaHost;
	LPHOSTENT   lpHostEntry;

	SOCKET      Socket;

	LPSERVENT   lpServEnt;
	SOCKADDR_IN saServer;

	int nRet;

	static char szBuffer[4096];

    long        start, end;
    long        bytes = 0;

	//
	// Use inet_addr( ) to determine if we're dealing with a name
	// or an address
	//
	iaHost.s_addr = inet_addr( lpServerName );
	if ( iaHost.s_addr == INADDR_NONE )
	{
		// Wasn't an IP address string, assume it is a name
		lpHostEntry = gethostbyname( lpServerName );
		if ( lpHostEntry == NULL )
		{
			PRINTERROR( "gethostbyname( )" );
			return;
		}
	}
	else
	{
		// It was a valid IP address string
		lpHostEntry = gethostbyaddr( (const char * )&iaHost,
									 sizeof( struct in_addr ), AF_INET );
		if ( lpHostEntry == NULL )
		{
			PRINTERROR( "gethostbyaddr( )" );
			return;
		}

		lpServerName = lpHostEntry->h_name;
	}


	//
	// Create a TCP/IP stream socket
	//
	Socket = socket( AF_INET, SOCK_STREAM, IPPROTO_TCP );
	if ( Socket == INVALID_SOCKET )
	{
		PRINTERROR( "socket( )" );
		return;
	}


	//
	// Find the port number for the HTTP service on TCP
	//
	lpServEnt = getservbyname( "http", "tcp" );
	if ( lpServEnt == NULL )
		saServer.sin_port = htons( 80 );
	else
		saServer.sin_port = lpServEnt->s_port;


	//
	// Fill in the rest of the server address structure
	//
	saServer.sin_family = AF_INET;
	saServer.sin_addr = *( (LPIN_ADDR )*lpHostEntry->h_addr_list );


	//
	// Connect the socket
	//
	fprintf( stderr,"\nConnecting to:%Fs\n", lpServerName );
	nRet = connect( Socket, ( LPSOCKADDR )&saServer, sizeof( SOCKADDR_IN ));
	if ( nRet == SOCKET_ERROR )
	{
		PRINTERROR( "connect( )" );
		closesocket( Socket );
		return;
	}


	//
	// Format the HTTP request
	//
	sprintf( szBuffer, "GET %Fs HTTP/1.1\n"\
					   "Host: %Fs\n"\
					   "Connection: close\n"\
					   "User-Agent: GetHTTP 0.0\n"\
					   "\n",
					   lpFileName, lpServerName );
	fprintf( stderr, "\nSending:\n%s", szBuffer );
	nRet = send( Socket, szBuffer, strlen( szBuffer ), 0 );
	if ( nRet == SOCKET_ERROR )
	{
		PRINTERROR( "send( )" );
		closesocket( Socket );
		return;
	}


	//
	// Receive the file contents and print to stdout
	//
	fprintf( stderr,"\nReceiving...\n" );

	start = clock();
	while ( 1 )
	{
		// Wait to receive, nRet = NumberOfBytesReceived
		nRet = recv( Socket, szBuffer, sizeof( szBuffer ), 0 );
		if ( nRet == SOCKET_ERROR )
		{
			PRINTERROR( "recv( )" );
			break;
		}

		// Did the server close the connection?
		if ( nRet == 0 )
			break;

		bytes += nRet;

        fprintf( stderr,"\nrecv( ) returned %d bytes:\n", nRet );

		// Write to stdout
		fwrite( szBuffer, nRet, 1, stdout );
	}
	end = clock();

    fprintf( stderr, "\nBytes:%ld Ticks:%d\n", bytes, end - start);

	//
	// Finish the connection
	//
	closesocket( Socket );
}

File: exs/pas/gethttp.pas
program
	gethttp;

(*
 * gethttp.pas
 *
 * Retrieves a file using the Hyper Text Transfer Protocol
 * and prints its contents to stdout.
 *)

(*
 * Pass the server name and full path of the file on the
 * command line and redirect the output to a file. The program
 * prints messages to stderr as it progresses.
 *
 * Example:
 *		GetHTTP www.idgbooks.com /index.html > index.html
 *)

uses
    dsock, strings;

const
	NL = #13#10;


(*************************************************************************)
procedure PRINTERROR( s: string );
begin
	writeln( NL, 'ERROR:', WSAGetLastError, ' @ ', s );
end;

(*************************************************************************)
procedure Get_HTTP( lpServerName, lpFileName: pchar );
var
	iaHost		: IN_ADDR;
	lpHostEntry	: LPHOSTENT;

	Sock		: TSocket;

	lpServEntry	: LPSERVENT;
	saServer	: sockaddr_in;

	nRet		: integer;

	szBuffer	: array[0..1023] of char;

begin
	(*
	 * Use inet_addr() to determine if we're dealing with a name
	 * or an address
	 *)
	iaHost.s_addr := longint( inet_addr( lpServerName ) );
	if ( iaHost.s_addr = INADDR_NONE ) then begin
		{ Wasn't an IP address string, assume it is a name }
		lpHostEntry := gethostbyname( lpServerName );
		if ( lpHostEntry = NULL ) then begin
			PRINTERROR( 'gethostbyname()' );
			exit;
		end;
	end
	else begin
		{ It was a valid IP address string }
		lpHostEntry := gethostbyaddr( pchar( iaHost ),
									  sizeof( in_addr ), AF_INET );
		if ( lpHostEntry = NULL ) then begin
			PRINTERROR( 'gethostbyaddr()' );
			exit;
		end;

		lpServerName := lpHostEntry^.h_name;
	end;


	(*
	 * Create a TCP/IP stream socket
	 *)
	Sock := socket( AF_INET, SOCK_STREAM, IPPROTO_TCP );
	if ( Sock = INVALID_SOCKET ) then begin
		PRINTERROR( 'socket()' );
		exit;
	end;


	(*
	 * Find the port number for the HTTP service on TCP
	 *)
	lpServEntry := getservbyname( 'http', 'tcp' );
	if ( lpServEntry = NULL ) then
		saServer.sin_port := htons( 80 )
	else
		saServer.sin_port := lpServEntry^.s_port;


	(*
	 * Fill in the rest of the server address structure
	 *)
	saServer.sin_family := AF_INET;
	saServer.sin_addr := lpHostEntry^.h_addr_list^^;


	(*
	 * Connect the socket
	 *)
	writeln( NL, 'Connecting to:', lpServerName );
	nRet := connect( Sock, sockaddr( saServer ), sizeof( sockaddr_in ) );
	if ( nRet = SOCKET_ERROR ) then begin
		PRINTERROR( 'connect()' );
		closesocket( Sock );
		exit;
	end;


	(*
	 * Format the HTTP request
	 *)
	strcopy( szBuffer, 'GET ' );
	strcat ( szBuffer, lpFileName );
	strcat ( szBuffer, ' HTTP/1.1' + NL + 'Host: ' );
	strcat ( szBuffer, lpServerName );
	strcat ( szBuffer, NL +
					   'Connection: close' + NL +
					   'User-Agent: GetHTTP 0.0' + NL +
					   NL );
	writeln( NL, 'Sending:', NL, szBuffer );
	nRet := send( Sock, szBuffer, strlen( szBuffer ), 0 );
	if ( nRet = SOCKET_ERROR ) then begin
		PRINTERROR( 'send()' );
		closesocket( Sock );
		exit;
	end;


	(*
	 * Receive the file contents and print to stdout
	 *)
	writeln( NL, 'Receiving...' );
	while ( true ) do begin
		{ Wait to receive, nRet = NumberOfBytesReceived }
		nRet := recv( Sock, szBuffer, sizeof( szBuffer ) - 1, 0 );
		if ( nRet = SOCKET_ERROR ) then begin
			PRINTERROR( 'recv()' );
			exit;
		end;

		{ Did the server close the connection? }
		if ( nRet = 0 ) then
			exit;

		writeln( NL, 'recv() returned ', nRet, ' bytes:' );

		{ Write to stdout }
		szBuffer[nRet] := #0;			{ add null-term }
		write( szBuffer );
	end;

	(*
	 * Finish the connection
	 *)
	closesocket( Sock );
end;

(*************************************************************************)
var
	wVersionRequested: word;
	_wsaData	: WSADATA;
	nRet		: integer;
    serverName,
    fileName	: array[0..128] of char;

begin

	(*
	 * Check arguments
	 *)
	if ( ParamCount <> 2 ) then begin
		writeln( NL, 'Syntax: GetHTTP ServerName FullPathName' );
		exit;
	end;

	(*
	 * Initialize WinSock.dll
	 *)
    wVersionRequested := MAKEWORD( 1,1 );
	nRet := WSAStartup( wVersionRequested, @_wsaData );
	if ( nRet <> 0 ) then begin
		writeln( NL, 'WSAStartup():', nRet );
		WSACleanup;
		exit;
	end;

	(*
	 * Check WinSock version
	 *)
	if ( _wsaData.wVersion <> wVersionRequested ) then begin
		writeln( NL, 'WinSock version not supported' );
		WSACleanup;
		exit;
	end;

	writeln( NL, 'WinSock ver:', _wsaData.wVersion shr 8, '.',
							     _wsaData.wVersion and $FF,
				 ' desc:', _wsaData.szDescription );

	(*
	 * Call Get_HTTP() to do all the work
	 *)
	Get_HTTP( strpcopy( serverName, ParamStr(1) ),
			  strpcopy( fileName  , ParamStr(2) ) );

	(*
	 * Release WinSock
	 *)
	write( NL, 'Finishing...' );
	WSACleanup;
	writeln( ' done.' );
end.

